core/src/main/java/org/apache/iceberg/PositionDeletesTable.java

```diff
@@ -132,6 +138,35 @@ private Schema calculateSchema() {
                 Types.StringType.get(),
                 MetadataColumns.FILE_PATH_COLUMN_DOC));
 
+    // Calculate used ids (for de-conflict)
+    Set<Integer> currentlyUsedIds =
+        Collections.unmodifiableSet(TypeUtil.indexById(Types.StructType.of(columns)).keySet());
+    Set<Integer> usedIds =
```

Comment:
@RussellSpitzer i found another issue.  In some case where I remove a column, i did not have those originally column ids in 'usedIds'.  So this adds it.

For this to work, I massaged the Schema API to take in directly a callback to reassign id.  I think it actually simpliesi the API a bit as it removes the concept of 'metadataIds', which was probably confusing anyway.

Reply:
Yep this also fixes the situation I was worried about.